Add a proper MCTristate type

peter-b opened the pull request:
The engine extensively used a `Boolean` type which was similar to `bool` but allowed a third `Mixed` state.  This was confusing.

This PR adds an `MCTristate` value type and makes `Boolean` an alias for `bool`.  It should fix lots and lots of Windows build warnings.

peter-b committed:
engine: Fix Boolean stored in single bit

The `Boolean` type used in the engine is actually a tristate, which
permits a `Mixed` value (value 2).  This patch replaces all instances
of `Boolean` that are explicitly stored in a single bit with `bool`.

peter-b committed:
engine: MClockmessages is a boolean value

Ensure that `MClockmessages` is only ever set to `true` or to `false`,
because `operator--` isn't defined for `bool`.

peter-b committed:
engine: Add MCTristate type for things that need to be true AND false

Buttons are currently allowed to be in a "mixed" highlight state.
Previously the `Boolean` type has been overloaded for this, requiring
it to be an integer type rather than `bool`.

This patch introduces a new `MCTristate` type and modifies every place
in the engine that handles tri-state values to use it, simplifying
some code in the exec interface and providing additional compile-time
protection.

peter-b committed:
engine: bool and uint8_t are not interchangeable

Don't assume that a pointer-to-`bool` and pointer-to-`uint8_t` are
safely aliasable.

livecodeali commented:
@livecode-vulcan review ok b3af266

livecode-vulcan commented:
💙 review by @livecodeali ok b3af2660c4f28c42f26f12e38942698b19e84063

livecode-vulcan commented:
😞 test failure b3af2660c4f28c42f26f12e38942698b19e84063

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1013)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1019)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1011)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1012)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1019)
* try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1033)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1010)

peter-b committed:
engine: Remove default initialiser for MCTristate::value

This isn't supported by the MSVC2010 compiler, but was irrelevant
anyway because `MCTristate` has a default constructor.

livecodeali commented:
@livecode-vulcan review ok b6dfb83

livecode-vulcan commented:
💙 review by @livecodeali ok b6dfb8360767c2b9d17026ddc4454b0e09c32a42

livecode-vulcan commented:
😎 test success b6dfb8360767c2b9d17026ddc4454b0e09c32a42

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1017)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1023)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1015)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1016)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1023)
* try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1037)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1014)
